src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractGenerateNuGetPackageCommand.cs

```diff
@@ -1,40 +1,151 @@
+                    _buildManager.AdviseUpdateSolutionEvents(this, out _solutionEventsCookie);
+                }
+            }
+        }
 
         protected override async Task<bool> TryHandleCommandAsync(IProjectTree node, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut)
         {
-            if (!ShouldHandle(node)) return false;
+            if (!ShouldHandle(node))
+            {
+                return false;
+            }
 
-            await _threadingService.SwitchToUIThread();
+            if (await IsReadyToBuildAsync().ConfigureAwait(false))
+            {
+                // Save documents before build.
+                var projectVsHierarchy = (IVsHierarchy)UnconfiguredProject.Services.HostObject;
+                int hr = _buildManager.SaveDocumentsBeforeBuild(projectVsHierarchy, (uint)VSConstants.VSITEMID.Root, 0 /*docCookie*/);
+                ErrorHandler.ThrowOnFailure(hr);
 
-            // TODO: Generate NuGet package.
+                // Enable generating package on build ("GeneratePackageOnBuild") for all projects being built.
+                _generatePackageOnBuildPropertyProvider.OverrideGeneratePackageOnBuild(true);
+
+                // Kick off the build.
+                uint dwFlags = (uint)(VSSOLNBUILDUPDATEFLAGS.SBF_SUPPRESS_SAVEBEFOREBUILD_QUERY | VSSOLNBUILDUPDATEFLAGS.SBF_OPERATION_BUILD);
+                hr = _buildManager.StartSimpleUpdateProjectConfiguration(projectVsHierarchy, null, null, dwFlags, 0, 0);
+                return ErrorHandler.Succeeded(hr);
+            }
 
             return true;
         }
+
+        #region IVsUpdateSolutionEvents members
+        public int UpdateSolution_Begin(ref int pfCancelUpdate)
+        {
+            return VSConstants.S_OK;
+        }
+
+        public int UpdateSolution_Done(int fSucceeded, int fModified, int fCancelCommand)
+        {
+            _generatePackageOnBuildPropertyProvider.OverrideGeneratePackageOnBuild(false);
+            return VSConstants.S_OK;
+        }
+
+        public int UpdateSolution_Cancel()
+        {
+            _generatePackageOnBuildPropertyProvider.OverrideGeneratePackageOnBuild(false);
+            return VSConstants.S_OK;
+        }
+
+        public int UpdateSolution_StartUpdate(ref int pfCancelUpdate)
+        {
+            return VSConstants.S_OK;
+        }
+
+        public int OnActiveProjectCfgChange(IVsHierarchy pIVsHierarchy)
+        {
+            return VSConstants.S_OK;
+        }
+        #endregion
+
+        #region IDisposable
+        public void Dispose()
+        {
+            var buildManager = _buildManager;
+            if (buildManager != null)
+            {
+                if (Interlocked.CompareExchange(ref _buildManager, null, buildManager) == buildManager)
+                {
+                    // Unregister solution build events.
+                    buildManager.UnadviseUpdateSolutionEvents(_solutionEventsCookie);
```

Comment:
I could not find this requirement in the API documentation. @davkean can you confirm?